Managing notebook metadata

jasongrout opened the pull request:
Fixes #5200 and #5235 

Here are a few user stories:

Jane opens and edits a notebook, and wants to save it back a single time in a 'finished' state, meaning she wants to capture all of the view information in the notebook document, such as collapse state, widget state, etc. She doesn't typically want to save this information.

Tom regularly wants to save a notebook without any outputs, in a minimal state, with no extra information. He'd like to do this from a single keyboard shortcut, preferably as the default save action.

Mary opens a notebook that has initial view state (such as collapsed/scrolled cells, etc.). She works in the notebook, changing this view state, running new cells, etc. She'd like to save her code and output changes, but not change the original initial view state (e.g., she's expanded some cells, edited them, but she wants the saved notebook to reflect the original collapsed state, whatever it was.)

Jake has been working in a notebook, and would like to save off a copy in a different file with all outputs deleted, but with the cell view state as he currently has it. He doesn't want to modify his current file.

To try out this branch on [binder](https://mybinder.org), follow this link: [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/jasongrout/jupyterlab/save-with-options?urlpath=lab-dev)

## TODO

- [x] Pull in whatever makes sense from #6009 (tests, for example?)
- [x] Put back in convenience things in the notebook tracker, like current cell, etc.
- [x] UX review

jasongrout commented:
Here's the current status:
<img width="325" alt="screen shot 2019-02-11 at 9 30 32 am" src="https://user-images.githubusercontent.com/192614/52581487-bb02f180-2ddf-11e9-86b9-d34ac75461cd.png">
<img width="512" alt="screen shot 2019-02-11 at 9 28 14 am" src="https://user-images.githubusercontent.com/192614/52581368-69f2fd80-2ddf-11e9-9699-52c80560704b.png">

The basic idea is that an extension can register a 'save option', along with label/help text and a default choice for the action. The action (what the user can choose in a dropdown) is one of:
* `save` - Save the current state to the notebook model
* `clear` - Clear the state from the notebook model
* `previous` - leave the current state of the model alone.
The default choices are given by the setting system, or if there is no setting for the option name, the default action provided in the option registration. The 'remember my choices' checkbox records the names and values to the user settings for the notebook.

Here's the basic idea, from the perspective of an extension developer:

```
  tracker.saveOptions.push({
    name: 'collapse-scroll-state',
    label: 'Cell collapsed and scrolled state',
    help:
      'The input and output collapsed state, as well as the output scrolled state for code cells',
    action: 'save',
    callback: (w: Notebook, name: string, action: SaveAction) => {
      switch (action) {
        case 'clear':
          NotebookActions.clearCollapseScrollState(w);
          break;

        case 'save':
          NotebookActions.persistCollapseScrollState(w);
          break;

        case 'previous':
        default:
      }
    }
  });
```

A number of questions come up:

1. I don't think there is a lot of precedent for having two save menu items. It seems that usually there is a single 'save', then it is 'export' or 'save as' that exposes options for saving. Should we follow a pattern like that?
2. For some options, like the view state, or the widget state, where the optional data is some *initial* state that isn't changed in course of running the notebook, all three options make sense, i.e., clear the state in the model, save the current state to be the new initial state, or just leave the initial state in the model alone. However, for other options such as whether to clear outputs, which is being changed continuously as the notebook is used, `previous` doesn't really make sense, since we don't keep the original outputs. For clearing outputs, it makes more sense to have a checkbox, i.e., either clear the outputs or not.
3. Is it surprising that changing an option, such as 'cell outputs', actually changes the notebook drastically (i.e., clears outputs)?
4. If we 'remember my choices', should that affect the default save (without the dialog)?

jasongrout commented:
ping @ellisonbg, @tgeorgeux

ellisonbg commented:
@ian-r-rose will know better, but I think the menu option should be capitalized as "Save Notebook with Options..."

jasongrout commented:
> will know better, but I think the menu option should be capitalized as "Save Notebook with Options..."

Fixed.

ellisonbg commented:
Also dropdown options should be capitalized in title case.

jasongrout commented:
> Also dropdown options should be capitalized in title case.

Thanks for the small corrections. The much bigger elephants in the room remain... :)

afshin commented:
I think one common pattern is:

* "Save"
* "Save As..." where this brings up a dialog that is collapsed by default and allows you to add a file name but has a caret that can be expanded with a bunch of options.

What do you think of employing this UX pattern?

jasongrout commented:
> What do you think of employing this UX pattern?

That is where I was going with my question 1, asking what others thought :). I think one common use case is people actually saving to the same file with various options. This would be easier if the save as dialog defaulted to saving over the current file, though that also might be confusing.

afshin commented:
I just took a quick look at my editors and the three I use most have multiple "Save" options. They all have "Save" and "Save As..." and in all three cases, the "Save As..." is collapsed but can be expanded. I think I've grown accustomed to this pattern.

* Sublime Text
  <img width="377" alt="screen shot 2019-02-11 at 18 20 18" src="https://user-images.githubusercontent.com/159529/52584343-b5ca9480-2e29-11e9-8335-afa21625b0fb.png">
* Visual Studio Code
  <img width="323" alt="screen shot 2019-02-11 at 18 16 01" src="https://user-images.githubusercontent.com/159529/52584197-5cfafc00-2e29-11e9-8a52-2bda42f6bc2f.png">
* TextMate
  <img width="366" alt="screen shot 2019-02-11 at 18 16 25" src="https://user-images.githubusercontent.com/159529/52584212-63897380-2e29-11e9-8399-a6813856a06b.png">

And here is a collapsed "Save As..."
<img width="449" alt="screen shot 2019-02-11 at 18 16 45" src="https://user-images.githubusercontent.com/159529/52584226-6c7a4500-2e29-11e9-8bf0-122f404361a3.png">
Compared with an expanded "Save As..."
<img width="852" alt="screen shot 2019-02-11 at 18 19 54" src="https://user-images.githubusercontent.com/159529/52584324-a8150f00-2e29-11e9-8af0-677512affd33.png">

afshin commented:
It's not a perfect analogy because our options are more specific than just showing a nice file browser in the dialog, but I think it's good as prior art.

jasongrout commented:
Another precedent: macOS Preview "Export" option, which *does* default to overwriting the current file:
<img width="380" alt="screen shot 2019-02-11 at 10 26 00 am" src="https://user-images.githubusercontent.com/192614/52584658-7c713500-2de7-11e9-803f-256cfa607d66.png">

jasongrout commented:
Another precedent: Word on macOS "Save As...", again defaults to overwriting the current file. The options button here pulls up an application-wide generic "saving options" button, which isn't about setting options for this save, but about options for how word saves things.
<img width="597" alt="screen shot 2019-02-11 at 10 27 45 am" src="https://user-images.githubusercontent.com/192614/52584803-e984ca80-2de7-11e9-8658-8cdc9f994247.png">
Here is the html "save as..." that shows some save-specific options:
<img width="598" alt="screen shot 2019-02-11 at 10 28 34 am" src="https://user-images.githubusercontent.com/192614/52584877-16d17880-2de8-11e9-85b5-3e98ebbbf3ee.png">

afshin commented:
Yeah, I agree that defaulting to overwriting is reasonable.

ellisonbg commented:
I think you can also simplify and align the dialog controls. I am thinking about something that has labels right aligned and the controls left aligned. This is described in Apple's HIG here:

https://developer.apple.com/design/human-interface-guidelines/macos/fields-and-labels/labels/

I think it would make sense for us to create a `Label` react component to encode this pattern.

The only question is where to put the longer descriptions. from the HIG I think the best place would be under the control, L aligned with the control above. The description you have for the cell output is redundant with the Label.

Finally, I think we are using a smaller dropdown size in the UI typically (see the toolbar usage).

ellisonbg commented:
And Labels should be title case.

ellisonbg commented:
@jasongrout let me know if you want me to post a mockup of the visual design feedback.

jasongrout commented:
@ellisonbg - thanks again for a detailed review of the dialog. First, I think it makes sense to address the architecture questions listed above, such as whether we should even have a separate "save with options" dialog at all, or whether it should be folded into a Save As... dialog. Second, do the three options presented make sense, or should we have flexibility for checkboxes (like for cell outputs).

ellisonbg commented:
Are the "remember my choice" settings per notebook or lab wide?

jasongrout commented:
Lab-wide for notebooks. They store in the lab notebook settings

ellisonbg commented:
My main concern about using "Save Notebook As..." for this is the discoverability. It isn't obvious that users would like to look there for these options. So I think I like the "Save N with Options..." approach.

jasongrout commented:
Another reason a "Save with Options" approach is nice from a dev standpoint is that it is enabled on a per-tracker basis, i.e., on a per-view basis. So it is immediately visible whether this particular view of a document offers options. If we instead folded it into a Save As dialog (with a per-tracker/per-view options section), you'd have to open the save as dialog to see whether the particular view provided options. I'm not sure how confusing this would be, but the "Save with Options" approach does allow you to expose that information to the user more quickly.

I'll add some user stories in the description to help focus the discussion.

jasongrout commented:
> My main concern about using "Save Notebook As..." for this is the discoverability. It isn't obvious that users would like to look there for these options.

I think Darian is saying that it *is* an established pattern to have options in the Save As... or Export... dialog.

In the original issue, we'd also have two new menu items, "Save Minimal" and "Save Maximal" or something, that minimize or maximize the information. Of course, one step up from that would be to have user-defined profiles for saving.

Another nice thing from a dev standpoint is that the current experiment is implemented solely in the notebook tracker. Folding it into the Save As… dialog means that we are committing to a much deeper change. Perhaps for now, in the experimentation phase, it makes sense to keep this a notebook-only thing?

What other notebook metadata might a user want control over in saving?

jasongrout commented:
One other much smaller note: in the current design, each option has a unique name that appears in the settings, but no other user-visible location, so it might be difficult for a user to map the options in the dialog to the settings in the Advanced Settings...

ivanov commented:
Something to consider: will this action change the underlying file or create a new file. I can imagine use cases for both.

`Save` - could save with default choices.
`Save as ...` - provides dialog to change name and notebook saving options, including a 'remember my choices', per @jasongrout's point 4 in [this comment](https://github.com/jupyterlab/jupyterlab/pull/5968#issuecomment-462422537). Right now, the "Save Notebook As..." dialog just allows you to change the filename. On the other hand, persisting this across notebooks withing a view may not make sense, and it would be weird to hit "Save" and have accidentally lost your output if you previously checked the "clear outputs" option and forgot about it, or thought it only applied to that notebook, as opposed to all notebooks.

`Export Notebook As ...` provides a "Notebook with options" target, which brings up the dialog and saves a copy with the options applied.

It seems like another option would be to have a notion of export options for Notebooks, and be able to apply those options to a notebook as a separate step. That way "Save" stays simple, and users like Tom "Apply Notebook Export Options" and save separately.

I think of the proposed user stories, Mary's is an outlier.  If we start considering the saving of the source of the cell without saving changes to the metadata, why not the reverse (save metadata changes only, without saving the source?) That just seems weird, because you're going down the path of keeping track of changes in every cell, for both metadata and source level changes, and then why not show all of these changes and allow users to save changed sources and metadata on a cell-by-cell basis, like `git add -p`.

jasongrout commented:
I think Mary's story is not so far-fetched - imagine if you have a notebook nicely laid out with collapsed introductory cells, etc., and Mary wants to add new cells at the end, etc. I can imagine she doesn't want to change the look of the notebook, just wants to update the code, and doesn't want to figure out all the view-level changes she made either.

On the other hand, if Mary had a "reset view state to initial state", to apply the existing initial view state metadata to her notebook, that would be useful in her case. She just reverts the collapse/scroll state to the initial state before saving. It's also useful to her to be able to preview exactly what she is saving.

jasongrout commented:
> if Mary had a "reset view state to initial state", to apply the existing initial view state metadata to her notebook view, that would be useful in her case

Developing this idea a bit further, I think this takes care of the confusing "previous" option I have in the current dialog. Instead of saving the previous value, just have an option to read the initial view state from the cell metadata and make the view reflect it. In other words, it would collapse/scroll cells to reflect the metadata. I could imagine a similar command for widgets, to load widget state from the notebook. Then each option instead becomes a checkbox, whether to save or not save that particular state.

jasongrout commented:
Actually, a checkbox isn't extremely clear, to say "check to save, uncheck to clear". You might imagine uncheck to mean "leave alone". Perhaps it should be radio buttons to "save" and "clear".

One thing I'd like to avoid is the jupyter notebook widget way of doing things, which is to have a separate command to "clear" the current state, then you have to save with that cleared state. People found that confusing.

jasongrout commented:
> persisting this across notebooks withing a view may not make sense, and it would be weird to hit "Save" and have accidentally lost your output if you previously checked the "clear outputs" option and forgot about it, or thought it only applied to that notebook, as opposed to all notebooks.

I think Paul is right here - it would be disconcerting to have 'clear outputs' be easily set as a default.

jasongrout commented:
Thinking more about Paul's thought-provoking comments, and the nature of the data we are talking about (i.e., view-level data which can be stored in the notebook model as initial view state), it's almost like what we really need is very clearly the concept in the user's mind of "save current view data to the notebook model as its initial view state" and "load saved initial view data from the model into the current view". Perhaps the view to model direction has both an "apply" and an "apply and save" button for convenience. And perhaps "clear outputs" is left out of this, since it already is a separate operation that is well-understood.

jasongrout commented:
@ivanov and I talked a lot more about this today, trying to tease out exactly what we want the user to have as a mental model, and how to make that the natural thing to conceive. We explored lots of ideas around how state is the same or different between views, and how it is shared with the model. Basically, we decided it was pretty complicated.

In the end, we came away with a much simpler proposal that we think will match user expectations and will provide an easy way to deal with the vast majority of usecases:

1. View state such as collapse and scroll state should *always* be saved in the notebook as initial state. Clearly the view you are saving from will determine this state.
2. Widgets should have a boolean setting, available in the settings menu, which is "Save widget state" (or something). If checked, the widget manager always serializes its state to the notebook. If not checked, the widget manager clears whatever state is in the notebook model on save. There won't be an option to preserve the existing widget state in the notebook - I think that's a fairly rare case.

This means that "save" should have some way of invoking view-specific callbacks.  In this PR currently, that's easy since the plugin is providing a view-specific menu item.    So one task is somehow exposing view-specific things in the current docmanager save command, or introducing a level of indirection in the save command.

Thoughts, @saulshanabrook (who implemented the current save-with-view-state command) or others?

As an aside, we realized that if widget state is really shared among all views, it should really be attached to the notebook model, rather than attached to the view. Perhaps an attached property on the model can make it easily accessible to all views.

jasongrout commented:
Some random thoughts and notes to self, captured to help me think more about this after looking at the code some more:

* Saving happens in the context: `context.save()`. There are lots of places this gets called, not just from the command (for example, 'save all', in trusting the notebook, etc.). Usually (always?) there is a widget available in the code that calls the context save. So if we want to have view methods called, we'll probably need to pass the view into the context save, or explicitly call the view-model sync before calling the context save.
* in a "save all", I suppose the view state of whatever view we pick to save will be saved with the latest idea of saving?

For widgets, again, perhaps the widget state really should be associated with the notebook model, and there register a callback with the notebook model that is called in the toJSON method of the notebook model, or perhaps the notebook model should have a signal that is emitted when it is about to serialize itself, which gives slots a chance to change the model before the serialization actually happens. 

*looks at code a bit more*

Actually, the context *already* has a signal, the `saveState` signal, that is emitted when a save is about to happen. So perhaps the widget manager can listen to this to figure out how to handle serialization. And perhaps each view could listen to this, and if it is the active widget, it could serialize its view state to the model too - this wouldn't handle the 'save all' case, but would handle saving when a widget view is active. Perhaps that is enough.

jasongrout commented:
> View state such as collapse and scroll state should _always_ be saved in the notebook as initial state. Clearly the view you are saving from will determine this state.

So with this, does changing the view state make the file dirty? If I collapse a cell, does the file instantly become dirty? If not, is there a shortcut and the save not done (since the file isn't 'dirty')?

saulshanabrook commented:
> * View state such as collapse and scroll state should _always_ be saved in the notebook as initial state. Clearly the view you are saving from will determine this state.

Do you mean that collapsed state is always saved in the file? What do you mean by "initial state"?

jasongrout commented:
> Do you mean that collapsed state is always saved in the file?

Yes, that was our thought.

> What do you mean by "initial state"?

Any collapsed/scrolled state in the model doesn't automatically change all views immediately. Instead, it becomes the initial view state when a view is created, right?

jasongrout commented:
In the dev meeting today, several people expressed preference for having something more like the dialog originally presented here, i.e., having explicit "save view state to model" functionality, rather than implicitly saving the view state by default.

jasongrout commented:
Thinking out loud again: How about instead of "Save with Options", a "Manage View State" dialog, or preference menu, or something. Basically, it presents each piece of view state, and asks for each one if you want to save the current state to the model, revert current state to the model's initial state, or automatically keep in sync. It's a bit more complicated, but starts to broach the ideas brought up by several people that in collaboration, you may want fine-grained sharing preferences.

jasongrout commented:
> The notebook instance tracker `activeCell` and `activeCellChanged` are definitely used in the wild, for what it's worth.

Oh. Do you know where they are used? I removed them in the latest commits, but can put them back if they are really used elsewhere. It seemed it was rather specific to the cell tools, so I wanted to trim down our API surface.

ian-r-rose commented:
They are used in the `jupyterlab-vim` and `jupyterlab-sublime` extensions, which use them to set vim and sublime keymaps on notebook cells, respectively.

ian-r-rose commented:
I'm less sure about whether the selection-related functionality is used in any extensions.

jasongrout commented:
I'm taking a look at jupyterlab-vim (https://github.com/jwkvam/jupyterlab-vim/blob/b297065dd874f00b213f95444613f2dbb815911b/src/index.ts). It seems that it uses the tracker activeCell signal and active cell in order to modify the codemirror for whatever cells the user actually goes to. It does make some assumptions that the editor is a codemirror instance. Perhaps it's better for that plugin to provide an editor factory to the system for notebooks? Then it will automatically be on all notebook cells, not just the cells that were active, and it will also know it is using codemirror.

jasongrout commented:
Looks like https://github.com/ryantam626/jupyterlab_sublime/blob/6813490d11d5df24ed453cb020885217294bed30/src/index.ts#L223 also uses `.activeCell` as a nice convenience, instead of `tracker.currentWidget.content.activeCell`, with the appropriate check for currentWidget, etc.

jasongrout commented:
Quick screenshot status update (clearly we need some ux work :)
<img width="312" alt="screen shot 2019-02-21 at 10 58 20 am" src="https://user-images.githubusercontent.com/192614/53194212-ac7abe00-35c7-11e9-8f1a-aeaecd619ba0.png">

saulshanabrook commented:
Sweet! What if we put the buttons next to the relevant item? If a button handles many of them, then maybe put a sub-header and put the button there?

jasongrout commented:
Next up is a setting for automatically syncing the collapse and scrolled data between the view and model (which means the model reflects the current state, and all views are lock-stop to it). How should we do this? Some options:

* Each cell has a changed signal, we listen to all cells everywhere, and make the appropriate model changes.
* Each cell sends a message to its parent notebook, the notebook merges this message stream, and we can subscribe at the notebook level.
* each cell has a flag for syncing view state to model state.

jasongrout commented:
> What if we put the buttons next to the relevant item? If a button handles many of them, then maybe put a sub-header and put the button there?

Can you sketch out what you have in your mind?

jasongrout commented:
CC @tgeorgeux - see the screenshot above - I don't think I'm *quite* to the point where a polished UX design is needed, but perhaps some rough ideas about what a design above would look like would help guide decisions.

Minimally, I think we should have two collapsible sections: Notebook-level things and cell-level things. I imagine the buttons syncing state and the notebook metadata are in the notebook-level section. I also imagine some of the true/false values in the cell section should be checkboxes.

jasongrout commented:
The latest changes add a notebook jlab setting for the cell metadata properties that will automatically be synced (defaulting to all of them: editable, collapse state, and scrolled state). I also removed all of the save with options menu and dialog things.

jasongrout commented:
Here's the latest state:

<img width="547" alt="screen shot 2019-02-23 at 12 45 20 pm" src="https://user-images.githubusercontent.com/192614/53291496-f8b13400-3768-11e9-80bb-398c72453b69.png">

Now I think I'm ready for some general review, including UX review. @tgeorgeux?

I think I'd still like to move the collapse panel to the general utilities, and there one other TODO note in the code.

To try out this branch on [binder](https://mybinder.org), follow this link: [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/jasongrout/jupyterlab/save-with-options?urlpath=lab-dev)

ellisonbg commented:
This is really challenging design problem, and it is wonderful you are diving into it and trying out different things. One thought is that it may be much easier and faster to mock this up in Figma (along with interactions) than exploring with code. I am working on code reviews today and available to talk if you want.

UX review:

* Last summer, the designers did a lot of work on the cell tools area, with the cell label extension. That extension began to move the cell tools extension to something that was used once in a while, to something that would see regular usage. My experience with it is that the tight coupling between a selected cell and the state of that UI was a key part of the good experience. There was a good mental model, signifiers, and mappings that made it easy to use without thinking. As a user navigated around JLab, that area would always show me what they needed and the information (slides or labels or metadata) was always in the same place. After playing with this branch, I feel that experience is broken in a manner than makes the cell tools extension good for neither notebooks or cells (rather than both). 
* Having the notebook metadata alone at the bottom feels like a minor issue (as in #6009), but as the area gets longer with more notebook content, it feels painful as it switches based both on cells and notebooks. I think the issue here that the inspector pattern (change what is shown based on selection) is a form of mildly modal interaction (modal in the sense of Raskin's Humane Interface https://en.wikipedia.org/wiki/The_Humane_Interface). If the user is aware of how selection is driving what is shown, it makes sense. However, if they miss that, it really is modal (hidden state controlling what the user experiences) and becomes very difficult to understand. 
* Based on this, I think the inspector UI for cells and notebooks should be separated out - either in two separate L sidebars, or as *possibly* two tabs in a single panel. The only problem with the later is that tabs in tabs is generally painful and a sign something is off.
* The mental model suggested by the buttons at the top is reasonably clear (state lives in two places, view + metadata, and the user can manually move state back and forth when they want), but the UX doesn't express that mental model, or behave in a manner that I can understand. When I change the scroll state or collapsed state of a cell, the cell metadata (in the L panel) changes immediately. Based on the above mental model, I was expecting to have to do that manually. I also don't see a setting to control this auto-sync behavior.
* Even though I think I understand the mental model, the new dropdowns in the cell area don't make any sense to me. Is this additional metadata? Where is it stored? Do I need to know about it?

jasongrout commented:
> One thought is that it may be much easier and faster to mock this up in Figma (along with interactions) than exploring with code.

I really wanted to explore it architecturally first, but I think now I'm to the point that I understand some of the basic underlying problems (both in terms of what I want and how to make it work), and so I think now is a good point to explore the interaction in figma.

jasongrout commented:
Let's have a discussion meeting to go over your other questions, and then summarize here. One thing: are you using the latest branch (i.e., does it look like the screenshot I posted in https://github.com/jupyterlab/jupyterlab/pull/5968#issuecomment-466692888 ?) You should be able to see the latest state by clicking on the "launch binder" button in the issue description. Your comments seem to indicate you might be looking at an older version of the UX exploration.

jasongrout commented:
@ellisonbg and I talked for a while about managing the notebook model and metadata. Here are our notes/proposal (though I think we may still have some disagreement on some details, or at least wording):

## Design principles for scalar view state

This state includes collapsed state, output scrolled state, editable state, etc.:

1. There is a model and a view
2. The model always agreed across multiple users and (eventually) with the file
3. The model and views can be synchronized or not
4. The initial view state is constructed from the model
5. Defer conflict resolution tools for conflict between model and view to the point where it is possible to have conflict (and to when the user should reasonably know what they prefer and have motivation to resolve the conflict?)
6. For 1.0, have all sync settings to true.

## Widgets

1. There is a kernel state, view model, view state, and file state
2. The kernel state, view model, and view state are always synchronized
3. If a file is opened with widget state, the view model and view state are hydrated from the file state
4. Identify relevant actions for managing state and create commands
5. Also offer a sync setting


## Notebook tools panel

We want common actions near the top, less common "advanced" actions toward the bottom and collapsed. It's tricky to know exactly what is "common" and what is "advanced". Raw editing of the metadata is advanced. Cell tags is probably common. Probably the user should have some say in what is considered common vs advanced?

jasongrout commented:
For now, @ellisonbg and I agree that we should simplify things drastically and just make the cell collapse state, cell editable state, and cell output scrolled state synced between the view and the model. This means that any change to any of these affects all views. This is a really simple model, with a very simple UX (e.g., checkboxes in the notebook tools panel, if you want something there). There are no buttons to transfer state to/from the model metadata.

As real-time collaboration becomes a reality, we will explore more fully how view state should be shared/synced, and how to make it user configurable.

So to people following this thread, how do you feel about making all view state described here as automatically synced to the model in 1.0? Note that this is also backwards-compatible with the classic notebook. What you see is what you get (in the model) is what is saved to disk.

ivanov commented:
> So to people following this thread, how do you feel about making all view state described here as automatically synced to the model in 1.0

Sounds good to me. WYSIWYS ("What you see is what you save") is easy to reason about. If you want to save something different than what you're seeing, you can always make a copy first and then change it to match whatever you'd like to save.

ellisonbg commented:
Thanks for posting the notes @jasongrout - I will be back around on Friday for another review round/discussion.

bollwyvl commented:
>
> What you see is what you get (in the model) is what is saved to disk.
>

Sounds good, almost always the right call. Checkboxes or switches sound
great, though short select boxes can be clearer for indicating mode changes
while being less pretty.

I could see lumping the persisted features as vertical tabs that you drag
between two "persistent" and "ephemeral" tab bar drop targets, but since
this is vertical, and the list could get long/change size, and we don't use
it anywhere else, it might be too much.

To that end, and not to hold up this effort: I don't like things popping in
and out of my my main app side bar buttons, or writing 30 char json haiku.
I see the metadata panel as persistent, and optionally expandable to a full
dock panel widget, with lots more space for editing, documentatiom, etc.
Could probably just be done with flexbox and some conditional styling.

Many other mime types also have metadata, like EXIF on images, an extra
file system in PDF, tolerances in cad documents. Making it a docregistry
feature would normalize all of this... And providing the data in a schema
(maybe with a react-schema-form UI schema) would mean a renderer wouldn't
even need to worry about many kinds of UI/validation/documentation drudgery.

Specifically for the Notebook: cells might even go nicely in a table. With
the loss of cell toolbar, it can be hard to lay out a slideshow, or manage
a bunch of different tags whereas in classic you can visually scan the
right margin. Namespaces would look great in grouped columns.

Surfacing metadata field documentation in inspector would also make a lot
of sense.
Another step towards making meta seem like less of an afterthought would be
to respect the Notebook metadata "title" in tab labels, showing cell
"name," etc.

jasongrout commented:
I simplified things a lot in the last few commits:

* There is only model-view synced collapse state, editable state, and output scrolled state. There no longer is an option to disable the syncing. I removed this option in a73737bd2e4cd99b08bee6c65b378b52b2d842da (so it is easy to go back and see how to add it again if needed).
* I moved the collapse widget to apputils, and made the json editor not have title/collapse functionality.
* I rearranged the notebook tools to address some thoughts Brian had about putting advanced tools behind a collapsed section:

<img width="477" alt="screen shot 2019-02-26 at 9 38 12 pm" src="https://user-images.githubusercontent.com/192614/53468696-194ee780-3a10-11e9-9654-85ce3acd94bd.png">

tgeorgeux commented:
https://www.figma.com/proto/SW6XIQEjWdtIjzxxZDWc8y2V/5698?node-id=3%3A51&viewport=774%2C-237%2C0.386558&scaling=scale-down

There's a simple prototype of what Brian and I spoke about this morning. I'm going to open a further issue to how we'll treat the 'cell tools' sidebar down the road. Below are the screenshots from the prototype for anybody who's having trouble loading the above:
![5698 - 1](https://user-images.githubusercontent.com/22109376/53922091-cc7a9a80-4027-11e9-874a-80991f038802.png)
![5698 - 2](https://user-images.githubusercontent.com/22109376/53922092-cc7a9a80-4027-11e9-9f78-a6d7792a32da.png)

tgeorgeux commented:
@ellisonbg we had a small bulleted list you'd typed out from this morning did that ever get posted?

ellisonbg commented:
@tgeorgeux hmmm, I did type it out into this thread, but it didn't get saved. Could you attempt to summarize the changes as they are in your screenshots?

jasongrout commented:
Thanks! So the standard place to put those collapse/expand arrows is on the left, not way over on the right. Is there a reason for putting them on the right?

Also, the metadata potentially is *very* large (for example, if there is widget state). That's why I collapsed the notebook metadata by default. If we're going to just display it without an option of collapsing it, we might want to put a max height on that notebook metadata editor?

ellisonbg commented:
@jasongrout - yeah these questions came up when @tgeorgeux and I talked.

* For the "expand icon" in this case, we wanted to keep the labels left aligned, and the dropdown's above already had a R aligned expand icon.
* I too am concerned about the large amount of content that can appear in the cell and notebook metadata editors. The solution that I proposed to Tim, was to have the entire cell tools switch to an alternate view - in other words, if a user wants to look at the advanced settings, hide the "normal" ones.

Something like this:

![screen shot 2019-03-06 at 11 41](https://user-images.githubusercontent.com/27600/53934026-d915e780-4055-11e9-8f1a-c634977f8caa.png)

What Tim and I have no data on is if users will really need to see the basic and advanced tools at the same time. I think they won't, but am not sure. Tim and I didn't spend any time on the interaction design of how a user switches between these two views. Some of the motivation for this design was my experience in using your original design that had lots of buttons, dropdowns and both metadata editors. The large scrollable content in the L sidebar began to feel really painful - and collapsing sections didn't help much.

jasongrout commented:
>     * For the "expand icon" in this case, we wanted to keep the labels left aligned, and the dropdown's above already had a R aligned expand icon.

I think a more conventional thing would be to put a small margin on the right for all of the labels.



> What Tim and I have no data on is if users will really need to see the basic and advanced tools at the same time. I think they won't, but am not sure.

I absolutely want to see both at the same time. I found having widgets controlling the metadata side by side with that metadata in raw form a very powerful transparent way to see exactly what was going on. For example, if you see both the cell metadata *and* the dropdown for the slide type, you see that changing the dropdown does nothing more than edit the metadata, and you see exactly how the metadata is edited. That natural understanding, by seeing the linked changes, was exactly what I wanted to form a mental model of how this information is stored and transferred.

If people can understand that all of those controls are affecting the metadata of a cell, and how they are affecting it (i.e., this control changes these metadata fields), I think that is a very powerful realization. That comes naturally if you see the metadata update in real-time as the controls are manipulated. If you don't see that update, it still feels magical and not clear at all.

In other words, part of the purpose here is to help people develop the mental model of where state lives, and how it is transferred between the server, the model, and the view. Seeing the raw data and recognizing it for what it is - the state in the model that is shared across all views, and will be serialized to disk, is a powerful natural effect of manipulating the controls and seeing the state update in real time.

jasongrout commented:
> For the "expand icon" in this case, we wanted to keep the labels left aligned, and the dropdown's above already had a R aligned expand icon.

Note also that the expand icon is in a section header, which already has a different font size and look from the other labels/controls. We could even emphasize the header look with a different background, border, or something. I don't think the section headers need to be aligned exactly with the content.

jasongrout commented:
@ellisonbg, @tgeorgeux - can we discuss this today?

tgeorgeux commented:
I am available to discuss today let me know what time works for you.

On Thu, Mar 7, 2019 at 8:04 AM Jason Grout <notifications@github.com> wrote:

> @ellisonbg <https://github.com/ellisonbg>, @tgeorgeux
> <https://github.com/tgeorgeux> - can we discuss this today?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyterlab/jupyterlab/pull/5968#issuecomment-470585481>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVFcwKGD0YudepkWXw6p0jMiTHEjQctzks5vUTiogaJpZM4axyWK>
> .
>

jasongrout commented:
@tgeorgeux and I iterated on things in a live session. Here's the current state:

<img width="421" alt="screen shot 2019-03-07 at 5 02 58 pm" src="https://user-images.githubusercontent.com/192614/54000096-ffaf4d80-413d-11e9-8fdc-55165cac0976.png">

jasongrout commented:
@tgeorgeux - I fixed that caret icon so it looks great now.

jasongrout commented:
@tgeorgeux, @ellisonbg - I just noticed that the extension manager has the same need for collapsible sections. In master, it uses the blueprint collapse control, which has the traditional caret on the left indicating the collapse state.

ellisonbg commented:
I am available to talk about this after noon today.

tgeorgeux commented:
I can meet between 12:30 and 1:30

jasongrout commented:
Adding some notebook tests, I realized we really do have a problem trying to reconcile conflicting metadata between the `collapsed` and `jupyter.outputs_hidden` metadata keys. I've opened https://github.com/jupyter/nbformat/issues/137 to discuss the appropriate thing to do.

In the meantime, I think perhaps that we ought to set the `jupyter.outputs_hidden` metadata as a side effect, but depend on the `collapsed` metadata key to be the source of truth for the state of the output rendering.

jasongrout commented:
I'm not seeing any conflicts here, but I merged in master anyway and pushed to try to satisfy github.

jasongrout commented:
I disagree with https://github.com/jupyterlab/jupyterlab/pull/3981#issuecomment-391139167. I think that until the ambiguity between `collapsed` and `jupyter.outputs_hidden` is clarified (see https://github.com/jupyter/nbformat/issues/137), or until `jupyter.outputs_hidden` becomes more popular as the way to indicate an output should be shown, we should just stick with reading/writing the `collapsed` metadata, as the classic notebook does. The problems with reading both fields is what to do when there is a conflict. The problem with writing both fields is that it could possible overwrite a conflict, without us recognizing there was a conflict. I think it's simpler to just not implement the (right now basically unused) `jupyter.outputs_hidden` field until the spec is clarified.

jasongrout commented:
(I amended the "Update our nbformat typings to understand source_hidden and outputs_hidden." commit with some minor formatting and comment modifications, leading to the force-push).

jasongrout commented:
I implemented the outputs_hidden strategy at https://github.com/jupyter/nbformat/issues/137#issuecomment-478167877 - basically anytime `jupyter.outputs_hidden` is set, it transfers the value to `collapsed`, or ignores it if `collapsed` is already set.

This does mean that setting `jupyter.outputs_hidden=true` (which then sets `collapsed=true`), then `jupyter.outputs_hidden=false` will ignore the `jupyter.outputs_hidden` (since `collapsed` is then set).

jasongrout commented:
Perhaps we should just convert the notebook cell metadata once when it is initially read from disk. Then the runtime behavior isn't confusing, the transformation is just done once when the data comes from disk.

jasongrout commented:
Okay, one final change here - at cell model creation, we make sure that `collapsed` and `jupyter.outputs_hidden` are synced (looking at `collapsed`, then `jupyter.outputs_hidden` in that order). From then on, at the cell model level, we make sure that these two pieces of metadata are synced.

jasongrout commented:
All checks passed! Thanks @afshin and @ian-r-rose and @saulshanabrook for your work on getting the test suite to pass!

afshin commented on packages/docmanager-extension/src/index.ts:
```diff
@@ -520,36 +520,21 @@ function addCommands(
+      return some(
+        map(shell.widgets('main'), w => docManager.contextForWidget(w)),
+        c => c && c.contentsModel && c.contentsModel.writable
+      );
```
🤓 
This is expressive and concise.

jasongrout replied:
higher-order iteration functions for the win! :)

jasongrout commented on packages/apputils/src/collapse.ts:
```diff
@@ -0,0 +1,160 @@
+    this.addClass('jp-Collapse');
+    this._header = new Widget();
+    this._header.addClass('jp-Collapse-header');
+    this._header.node.addEventListener('click', this);
```
No, I agree this should change. Good catch.

jasongrout replied:
Fixed.

jasongrout commented on packages/apputils/src/collapse.ts:
```diff
@@ -0,0 +1,160 @@
+    if (this.isDisposed) {
+      return;
+    }
+    super.dispose();
```
Yep, I have a mental note to look through my code to see where I called dispose too early.

jasongrout replied:
Fixed.

jasongrout commented on packages/apputils/src/collapse.ts:
```diff
@@ -0,0 +1,160 @@
+    // Delete references we explicitly hold to other widgets.
+    delete this._header;
+    delete this._widget;
+    delete this._content;
```
I've been trying to figure out how to deal with eventually turning on nullable checks without defining every parameter as ` T | null`. Perhaps it's better to go ahead and do `T | null` as the type for such attributes?

afshin replied:
I think `T | null` is not bad. I'm not sure either, though.

jasongrout replied:
`T | null` means that you always have to typecheck it when using it. That can be really annoying, even if it may be technically correct.

jasongrout commented on packages/apputils/src/instancetracker.ts:
```diff
@@ -461,8 +461,8 @@ export class InstanceTracker<T extends Widget>
         null;
-      this._currentChanged.emit(this._currentWidget);
       this.onCurrentChanged(this._currentWidget);
+      this._currentChanged.emit(this._currentWidget);
```
For the record, this is about making it consistent with the same emission logic in the previous function, not necessarily making a conscious decision about whether we should emit or call the handler first.

afshin commented on packages/cells/src/model.ts:
```diff
@@ -492,6 +493,36 @@ export class CodeCellModel extends CellModel implements ICodeCellModel {
+      Private.collapseChanged(this.metadata, {
+        type: 'change',
+        key: 'collapsed',
+        oldValue: collapsed,
```
Should the `oldValue` be different?

jasongrout replied:
I think technically it is correct, and it doesn't matter anyway in this particular invocation.

afshin commented on packages/cells/src/model.ts:
```diff
@@ -492,6 +493,36 @@ export class CodeCellModel extends CellModel implements ICodeCellModel {
+        Private.collapseChanged(this.metadata, {
+          type: 'change',
+          key: 'jupyter',
+          oldValue: jupyter,
```
Also here: should the `oldValue` be different?

jasongrout replied:
It can be, I suppose. Since it's a dummy message, technically I think it is correct - there is no change in the value.

afshin commented on packages/cells/src/model.ts:
```diff
@@ -620,3 +651,33 @@ export namespace CodeCellModel {
+      }
+    }
+  }
+}
```
I think there may be something wrong in this logic. Here is the behavior I am seeing:

1. Set a cell's metadata to `"collapsed": true` and see `outputs_hidden` get created.
1. Set `"collapsed": false` and save and it comes back again as `true`
1. Set `"collapsed": false` again and metadata block disappears.

![metadata](https://user-images.githubusercontent.com/159529/55577477-93f3dc80-570b-11e9-8d1d-5fec4c6f1f75.gif)

afshin replied:
When I am in the cell metadata editor, if I hit `Accel S` the notebook file saves, but the actually metadata I am trying to save doesn't save unless I click on the checkbox button. Do you think that should be fixed in a separate PR? I'll open an issue if so.

jasongrout replied:
Thanks - I agree it's a separate issue.

afshin replied:
Cool.

Is the behavior in the screen cap happening for you as well?

jasongrout replied:
> Is the behavior in the screen cap happening for you as well?

I haven't checked yet.

jasongrout replied:
Yes, I can reproduce the issue you saw. Investigating...

jasongrout replied:
Turns out that the metadata editor was setting *all* fields, instead of just setting fields that changed. That meant that when it set `collapsed`, the `outputs_hidden` was being updated properly, but then the metadata editor would set `outputs_hidden` to its old value, which would cause collapsed to be changed. I fixed the metadata editor to only set values that had actually changed from the original value.

tgeorgeux commented:
@jasongrout this is super nit-pickey, but I think if we could solve it now it would help in the future as well. 

I think the amount of padding or margin on the bottom should match the top. I tried to mess around with the CSS in the inspector, but I caused more problems than I fixed.
![Margin or padding](https://user-images.githubusercontent.com/22109376/55580184-cfbc8100-56ce-11e9-988f-a5c93eecd360.png)

jasongrout commented:
@tgeorgeux - thanks. Can you indicate on the image *exactly* what spaces should be equal?

tgeorgeux commented:
@jasongrout yes, sorry that picture wasn't as clear as I originally thought.

![more precise](https://user-images.githubusercontent.com/22109376/55583425-aef82980-56d6-11e9-8763-7bf703b6f125.png)

jasongrout commented:
@tgeorgeux I got the padding close:
![Screen Shot 2019-04-05 at 3 24 32 PM](https://user-images.githubusercontent.com/192614/55659526-a34d5580-57f1-11e9-9454-6abf9d9b7507.png)

jasongrout commented:
@afshin - it works better now: 
![screenshot](https://user-images.githubusercontent.com/192614/55659590-ead3e180-57f1-11e9-9a01-2a67b16637ae.gif)

jasongrout commented:
Yeah! Will you merge it?

afshin commented:
Boom!

meeseeksmachine commented:
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/could-not-find-model-due-to-outputs-text-plain-truncated-at-111-bytes/733/5

meeseeksmachine commented:
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/hiding-code-cell-on-launch/1763/5
